client/build_linux_test.go

```diff
@@ -0,0 +1,260 @@
+	"syscall"
+	"testing"
+
+	"github.com/containerd/console"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/solver/errdefs"
+	"github.com/moby/buildkit/solver/pb"
+	"github.com/moby/buildkit/util/testutil/integration"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func TestClientGatewayIntegrationLinux(t *testing.T) {
+	integration.Run(t, []integration.Test{
+		testClientGatewayContainerPID1Tty,
+		testClientGatewayContainerExecTty,
+	}, integration.WithMirroredImages(integration.OfficialImages("busybox:latest")))
+}
+
+// testClientGatewayContainerPID1Tty is testing that we can get a tty via
+// a container pid1, executor.Run
+func testClientGatewayContainerPID1Tty(t *testing.T, sb integration.Sandbox) {
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	product := "buildkit_test"
+
+	ptm, ptsName, err := console.NewPty()
+	require.NoError(t, err)
+	pts, err := os.OpenFile(ptsName, os.O_RDWR|syscall.O_NOCTTY, 0)
+	require.NoError(t, err)
+	err = ptm.DisableEcho()
+	require.NoError(t, err)
+
+	scanner := bufio.NewReader(ptm)
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		st := llb.Image("busybox:latest")
+
+		def, err := st.Marshal(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal state")
+		}
+
+		r, err := c.Solve(ctx, client.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to solve")
+		}
+
+		ctr, err := c.NewContainer(ctx, client.NewContainerRequest{
+			Mounts: []client.Mount{{
+				Dest:      "/",
+				MountType: pb.MountType_BIND,
+				Ref:       r.Ref,
+			}},
+		})
+		require.NoError(t, err)
+		defer ctr.Release(ctx)
+
+		pid1, err := ctr.Start(ctx, client.StartRequest{
+			Args:   []string{"sh"},
+			Tty:    true,
+			Stdin:  pts,
+			Stdout: pts,
+			Stderr: pts,
+		})
+		require.NoError(t, err)
+
+		// wait for prompt
+		line, err := scanner.ReadString('#')
+		require.NoError(t, err)
+		require.Equal(t, "/ #", line)
+
+		// cd && wait for prompt
+		ptm.Write([]byte("cd /tmp\n"))
+		line, err = scanner.ReadString('#')
+		require.NoError(t, err)
+		require.Equal(t, " cd /tmp\r\r\n/tmp #", line)
+
+		// create file
+		ptm.Write([]byte("echo foobar > newfile\n"))
+		line, err = scanner.ReadString('#')
+		require.NoError(t, err)
+		require.Equal(t, " echo foobar > newfile\r\r\n/tmp #", line)
+
+		// read file
+		ptm.Write([]byte("cat /tmp/newfile\n"))
+		line, err = scanner.ReadString('#')
+		require.NoError(t, err)
+		require.Equal(t, " cat /tmp/newfile\r\r\nfoobar\r\r\n/tmp #", line)
+
+		// test resize, note we cannot verify the console size in the container
+		// because we need a controlling terminal to report size back, and the
+		// simple ptm we are using is not enough.  Ideally we could just use
+		// `/bin/ttysize` but that reports `0 0` with this test.
```

Comment:
what's the reason for it reporting `0 0`? Afaics we don't need `NewPty` at all for this test, could just use `io.Pipe`.

Reply:
Not sure why ttysize does not work, I thought an actual terminal was necessary for this, but I might be wrong.  I will try it again after the updates, also will try with `io.Pipe()`

Reply:
updated the tests, removed the `NewPty` usage.  `ttysize` works now, the problem was I was sending the SIGWINCH to the pid from the pidfile but that was `Meta.Args` program (ie `sh`) and the resize only takes effect when we send SIGWINCH to the `runc run` process.  

It was a bit fussy getting the test reliable, there were spurious control characters printed at random times, not sure where they are coming from but I only noticed them after `ttysize` was added, not sure if is related.  To make the test more reliable we should probably use `github.com/hinshun/vt10x` or `github.com/Netflix/go-expect`, but those will bring in a few deps.  I figure we can try it as is, the short sleeps should keeps things consistent, but if it becomes unreliable we can disable or rework.

Reply:
One way to clean this up would be to run `stty -echo` as the first command. Then you don't need to worry about the timings between prompt echo and response strings. This leaves the `500ms` timeouts that waits for the resize. I guess these are fine but you could also write a helper that tests the buffer contents for expected contents every 50-100ms and has a longer 5-10sec timeout that produces failure if no content appears.